### PR TITLE
mkinitcpio: Add support for the zstd compressor

### PR DESCRIPTION
**Pull request by luca020400**
This mainly for testing purposes until https://lkml.org/lkml/2020/7/23/1010 is merged in mainline.
Then this patch can  be safely applied

### CODE REVIEW

File: mkinitcpio
Comment (luca020400) [inline mkinitcpio:215]: the options used are the same as found in the kernel proposal for the vmlinux image

**Comment by luca020400** [comment]
For anyone interested testing
I backported the patches here: https://github.com/luca020400/linux
Apply this diff to the config https://gist.github.com/6631c82f4ed0c76267a1288376c34f71

**Comment by luca020400** [comment]
@grazzolini Mind taking a look?

I uploaded the package to aur as mknitcpio-zstd ( along a supported kernel linux-llvm )

**Comment by grazzolini** [comment]
Hi @luca020400. I plan on adding zstd support to mkinitcpio as soon as the kernel supports it. I also planned on dropping lz4 support, since it's the only compression algorithm that's not reproducible. If the Arch kernel got patched to support zstd, I might consider patching the mkinitcpio package as well, but that's not something I'm considering pushing to the kernel maintainers currently.

**Comment by luca020400** [comment]
Feel free to use these commits, they've been tested.
zstd support might come in 5.9, so I guess we can wait.
As seen here https://lkml.org/lkml/2020/7/24/677 bzip2 and lzma are likely going to be dropped from kernel, so we might want to remove them as well.
Didn't know lz4 isn't reproducible, but honestly lz4 is a good algorithm extensively used thanks to the quick decompression, I've been using it before switching to zstd.
I think as soon as zstd lands lz4 removal is doable, and actually switching the default

**Comment by grazzolini** [comment]
I was told that at least one of the officially supported kernels for Arch, linux-zen, has zstd support, due to the fact that the zen patchset adds zstd support. I'll give your patches some testing and try to merge them and make a mkinitcpio release before 5.9, that is, considering 5.9 really adds zstd support.

**Comment by luca020400** [comment]
I hope it does, but all I can do is adding by "Tested-by" sign-off on the changes. I'd really want to see zstd being in.
Didn't know linux-zen support zstd compression, then I think you can safely add this feature even now.

**Comment by falconindy** [comment]
> then I think you can safely add this feature even now.
We have no checks in mkinitcpio for examining compression support. If a user on a non-zen kernel tries to produce a zstd initramfs, they're going to have a bad time.

I *really* don't see any rush for getting this into mkinitcpio mainline. Nothing about the current codebase prevents you (an interested early adopter) from using zstd on your own kernel.

**Comment by luca020400** [comment]
@falconindy agree, then we shall wait for mainline adoption of zstd

**Comment by grazzolini** [comment]
Well, it *can* land on mkinitcpio before that. I'm particular ok with adding zstd support but not adding it as default on mkinitcpio.conf. Perhaps not even list it there for now. At least one official kernel (-zen) supports it already. When zstd lands on mainline, we can add zstd to the default .conf file. And, after a lot of testing (and giving people time and warning), perhaps even change it to default.

**Comment by luca020400** [comment]
@grazzolini feel free to amend my commit to remove it from the default .conf if you want to do that.
I'd be happy :)

**Comment by eworm-de** [comment]
You should update the info in man page... Support for zstd was not added in linux 2.6.38.

File: mkinitcpio
Comment (grazzolini) [reply mkinitcpio:215]: You have a link for this?

Comment (luca020400) [reply mkinitcpio:215]: https://lkml.org/lkml/2020/7/23/1013
But I've actually read the documentation and -22 shouldn't be used, instead for initramfs -19 should do the job

**Comment by luca020400** [comment]
Incorporated changes made by @eworm-de for man page
Also fixed compression options to use -19

File: mkinitcpio.conf
Comment (grazzolini) [inline mkinitcpio.conf:63]: I'm still wary of adding this to the .conf file, even commented, because the users might get the impression that we support zstd, which we don't yet. I was thinking about reaching to the Arch kernel maintainers, @heftig, @anthraxx, and others, to see if we could have the zstd patch merged across the board, so this wouldn't be an issue.

**Comment by eworm-de** [comment]
Linux 5.8 hast just been released, linux 5.10 will be in about four month. That is when linux 5.9 moves to linux-lts package. So that is latest date to release full support.

File: mkinitcpio
Comment (codicodi) [inline mkinitcpio:215]: How about  `-T0` (multithreaded compression)?

Comment (luca020400) [reply mkinitcpio:215]: None of the tools used in mkinitcpio are threaded, so probably not

Comment (grazzolini) [reply mkinitcpio:215]: If -T0 is faster (should be) and still makes reproducible archives, that's great. On the other hand, I'd say we should use reasonable conservative defaults (matching upstream zstd defaults), and let users tweak further, if they want.

Comment (codicodi) [reply mkinitcpio:215]: It should be reproducible [1]. `-T0` is also used in devtools [2], but admittedly, unlike some packages, initcpios aren't usually large enough for parallel compression to make a big difference.

[1] https://github.com/facebook/zstd/issues/2238
[2] https://github.com/archlinux/devtools/blob/20200407/makepkg-x86_64.conf#L135

Comment (Maryse47) [reply mkinitcpio:215]: > I'd say we should use reasonable conservative defaults (matching upstream zstd defaults)

I think `-19` isn't upstream default already otherwise you wouldn't need to pass it explicitly.

Comment (klausenbusk) [reply mkinitcpio:215]: Debian uses `-19 -T0` for what it's worth: https://salsa.debian.org/kernel-team/initramfs-tools/-/blob/cbbbb1b73e72124ba517325c51e0ea4bd7139104/mkinitramfs#L187

Comment (luca020400) [reply mkinitcpio:215]: They also use pigz for gzip is available, we don't.
Maybe we can just do another PR to switch to multi-threaded tools if possibile?

Comment (grazzolini) [reply mkinitcpio:215]: I'm fine with using -T0 right now, as long as it's reproducible (it seems to be). Also, what I meant by -19 upstream is the kernel recommendation to use -19, not zstd's. Anyway, @luca020400, if you want to update this branch to add -T0, I'm fine with it.

Comment (luca020400) [reply mkinitcpio]: Done

Comment (heftig) [reply mkinitcpio]: Will it even matter in this case? What's the block size used for -19 compared to the size of the typical initramfs?

Comment (codicodi) [reply mkinitcpio]: I think you mean window size? It's 21M. Fallback initramfs can easily be 100+. 
On my machine `-T0` speeds up `mkinitcpio -P` by 35%.

Comment (grazzolini) [reply mkinitcpio]: The regular initramfs is around that window, but I'd say it still has benefits, even if you, like me, aren't using the fallback initramfs (I'm thinking about disabling it's generation *by default*)

**Comment by luca020400** [comment]
Seems like @heftig defaulted 5.9 to ZSTD ( which also brings in the CONFIG_RD_ZSTD needed for ramdisk )
Zen already backported them, I guess we have to wait for LTS now

**Comment by grazzolini** [comment]
I won't wait until LTS gets support for this. Now that 5.9 has zstd support, I'll merge this PR and make a new mkinitcpio release (after a ton of testing). If someone tries to use zstd with LTS, that's their fault, because the documentation will state accurately which kernel got zstd support. Nor mkinitcpio will switch to zstd by default, in the foreseeable future.

**Comment by luca020400** [comment]
Good for me then. Ship it

**Comment by eworm-de** [comment]
Greg fooled us... Linux 5.9 is not going to be LTS - but 5.10 will. So we will have to wait for two more month for zstd compressed initcpio with linux-lts.

**Comment by grazzolini** [comment]
Well, I've done a few tests already with 5.9, it works fine. I'm cleaning PR's this week and making a release. But anyone willing to use zstd can do so already, only lsinitcpio won't (properly) work on the images.

**Review by grazzolini — APPROVED** [review]

**Pull request merged**
